// larkfinserv-react-native-sdk/src/components/LarkFinServWebView.tsx
import React, { useEffect, useRef, useState } from 'react';
import {
  View,
  StyleSheet,
  ActivityIndicator,
  Modal,
  TouchableOpacity,
  Text,
  SafeAreaView,
} from 'react-native';
import { WebView } from 'react-native-webview';
import LarkFinServSDK from '../LarkFinServSDK';
import { SDKMode } from '../types';

interface LarkFinServWebViewProps {
  sdk: LarkFinServSDK;
  mode?: SDKMode;
  onClose?: () => void;
  visible?: boolean;
  containerStyle?: any;
}

const LarkFinServWebView: React.FC<LarkFinServWebViewProps> = ({
  sdk,
  mode = 'inline',
  onClose,
  visible = true,
  containerStyle,
}) => {
  const webViewRef = useRef<WebView>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  const iframeUrl = sdk.getIframeUrl();

  useEffect(() => {
    if (!iframeUrl) {
      setError('SDK not initialized. Please call sdk.initialize() first.');
    }
  }, [iframeUrl]);

  const handleMessage = (event: any) => {
    const message = event.nativeEvent.data;
    sdk.handleWebViewMessage(message);
  };

  const handleLoadEnd = () => {
    setIsLoading(false);
  };

  const handleError = () => {
    setIsLoading(false);
    setError('Failed to load SDK');
  };

  const handleClose = () => {
    if (onClose) {
      onClose();
    }
  };

  const renderWebView = () => (
    <View style={[styles.container, containerStyle]}>
      {isLoading && (
        <View style={styles.loadingContainer}>
          <ActivityIndicator size="large" color="#16a34a" />
          <Text style={styles.loadingText}>Loading SDK...</Text>
        </View>
      )}
      {error && (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{error}</Text>
        </View>
      )}
      {iframeUrl && !error && (
        <WebView
          ref={webViewRef}
          source={{ uri: iframeUrl }}
          style={styles.webview}
          onMessage={handleMessage}
          onLoadEnd={handleLoadEnd}
          onError={handleError}
          javaScriptEnabled={true}
          domStorageEnabled={true}
          startInLoadingState={true}
          scalesPageToFit={true}
          allowsInlineMediaPlayback={true}
          mediaPlaybackRequiresUserAction={false}
        />
      )}
    </View>
  );

  if (mode === 'inline') {
    return renderWebView();
  }

  // For embedded mode with modal
  return (
    <Modal
      visible={visible}
      animationType="slide"
      presentationStyle="fullScreen"
      onRequestClose={handleClose}
    >
      <SafeAreaView style={styles.modalContainer}>
        <View style={styles.header}>
          <TouchableOpacity onPress={handleClose} style={styles.closeButton}>
            <Text style={styles.closeButtonText}>✕</Text>
          </TouchableOpacity>
        </View>
        {renderWebView()}
      </SafeAreaView>
    </Modal>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  modalContainer: {
    flex: 1,
    backgroundColor: '#fff',
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    padding: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  closeButton: {
    padding: 10,
  },
  closeButtonText: {
    fontSize: 24,
    color: '#666',
  },
  webview: {
    flex: 1,
  },
  loadingContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
    zIndex: 1,
  },
  loadingText: {
    marginTop: 10,
    fontSize: 16,
    color: '#666',
  },
  errorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  errorText: {
    fontSize: 16,
    color: '#dc2626',
    textAlign: 'center',
  },
});

export default LarkFinServWebView;
