// larkfinserv-react-native-sdk/lib/module/components/LarkFinServWebView.js
import React, { useEffect, useRef, useState } from 'react';
import { View, StyleSheet, ActivityIndicator, Modal, TouchableOpacity, Text, SafeAreaView } from 'react-native';
import { WebView } from 'react-native-webview';
const LarkFinServWebView = ({
  sdk,
  mode = 'inline',
  onClose,
  visible = true,
  containerStyle
}) => {
  const webViewRef = useRef(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);
  const iframeUrl = sdk.getIframeUrl();
  useEffect(() => {
    if (!iframeUrl) {
      setError('SDK not initialized. Please call sdk.initialize() first.');
    }
  }, [iframeUrl]);
  const handleMessage = event => {
    const message = event.nativeEvent.data;
    sdk.handleWebViewMessage(message);
  };
  const handleLoadEnd = () => {
    setIsLoading(false);
  };
  const handleError = () => {
    setIsLoading(false);
    setError('Failed to load SDK');
  };
  const handleClose = () => {
    if (onClose) {
      onClose();
    }
  };
  const renderWebView = () => /*#__PURE__*/React.createElement(View, {
    style: [styles.container, containerStyle]
  }, isLoading && /*#__PURE__*/React.createElement(View, {
    style: styles.loadingContainer
  }, /*#__PURE__*/React.createElement(ActivityIndicator, {
    size: "large",
    color: "#16a34a"
  }), /*#__PURE__*/React.createElement(Text, {
    style: styles.loadingText
  }, "Loading SDK...")), error && /*#__PURE__*/React.createElement(View, {
    style: styles.errorContainer
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.errorText
  }, error)), iframeUrl && !error && /*#__PURE__*/React.createElement(WebView, {
    ref: webViewRef,
    source: {
      uri: iframeUrl
    },
    style: styles.webview,
    onMessage: handleMessage,
    onLoadEnd: handleLoadEnd,
    onError: handleError,
    javaScriptEnabled: true,
    domStorageEnabled: true,
    startInLoadingState: true,
    scalesPageToFit: true,
    allowsInlineMediaPlayback: true,
    mediaPlaybackRequiresUserAction: false
  }));
  if (mode === 'inline') {
    return renderWebView();
  }

  // For embedded mode with modal
  return /*#__PURE__*/React.createElement(Modal, {
    visible: visible,
    animationType: "slide",
    presentationStyle: "fullScreen",
    onRequestClose: handleClose
  }, /*#__PURE__*/React.createElement(SafeAreaView, {
    style: styles.modalContainer
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.header
  }, /*#__PURE__*/React.createElement(TouchableOpacity, {
    onPress: handleClose,
    style: styles.closeButton
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.closeButtonText
  }, "\u2715"))), renderWebView()));
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff'
  },
  modalContainer: {
    flex: 1,
    backgroundColor: '#fff'
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    padding: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#eee'
  },
  closeButton: {
    padding: 10
  },
  closeButtonText: {
    fontSize: 24,
    color: '#666'
  },
  webview: {
    flex: 1
  },
  loadingContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
    zIndex: 1
  },
  loadingText: {
    marginTop: 10,
    fontSize: 16,
    color: '#666'
  },
  errorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20
  },
  errorText: {
    fontSize: 16,
    color: '#dc2626',
    textAlign: 'center'
  }
});
export default LarkFinServWebView;
