// larkfinserv-react-native-sdk/lib/commonjs/components/LarkFinServWebView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeWebview = require("react-native-webview");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const LarkFinServWebView = ({
  sdk,
  mode = 'inline',
  onClose,
  visible = true,
  containerStyle
}) => {
  const webViewRef = (0, _react.useRef)(null);
  const [isLoading, setIsLoading] = (0, _react.useState)(true);
  const [error, setError] = (0, _react.useState)(null);
  const iframeUrl = sdk.getIframeUrl();
  (0, _react.useEffect)(() => {
    if (!iframeUrl) {
      setError('SDK not initialized. Please call sdk.initialize() first.');
    }
  }, [iframeUrl]);
  const handleMessage = event => {
    const message = event.nativeEvent.data;
    sdk.handleWebViewMessage(message);
  };
  const handleLoadEnd = () => {
    setIsLoading(false);
  };
  const handleError = () => {
    setIsLoading(false);
    setError('Failed to load SDK');
  };
  const handleClose = () => {
    if (onClose) {
      onClose();
    }
  };
  const renderWebView = () => /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, containerStyle]
  }, isLoading && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.loadingContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.ActivityIndicator, {
    size: "large",
    color: "#16a34a"
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.loadingText
  }, "Loading SDK...")), error && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.errorContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.errorText
  }, error)), iframeUrl && !error && /*#__PURE__*/_react.default.createElement(_reactNativeWebview.WebView, {
    ref: webViewRef,
    source: {
      uri: iframeUrl
    },
    style: styles.webview,
    onMessage: handleMessage,
    onLoadEnd: handleLoadEnd,
    onError: handleError,
    javaScriptEnabled: true,
    domStorageEnabled: true,
    startInLoadingState: true,
    scalesPageToFit: true,
    allowsInlineMediaPlayback: true,
    mediaPlaybackRequiresUserAction: false
  }));
  if (mode === 'inline') {
    return renderWebView();
  }

  // For embedded mode with modal
  return /*#__PURE__*/_react.default.createElement(_reactNative.Modal, {
    visible: visible,
    animationType: "slide",
    presentationStyle: "fullScreen",
    onRequestClose: handleClose
  }, /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: styles.modalContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.header
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: handleClose,
    style: styles.closeButton
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.closeButtonText
  }, "\u2715"))), renderWebView()));
};
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff'
  },
  modalContainer: {
    flex: 1,
    backgroundColor: '#fff'
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    padding: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#eee'
  },
  closeButton: {
    padding: 10
  },
  closeButtonText: {
    fontSize: 24,
    color: '#666'
  },
  webview: {
    flex: 1
  },
  loadingContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
    zIndex: 1
  },
  loadingText: {
    marginTop: 10,
    fontSize: 16,
    color: '#666'
  },
  errorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20
  },
  errorText: {
    fontSize: 16,
    color: '#dc2626',
    textAlign: 'center'
  }
});
var _default = exports.default = LarkFinServWebView;
