// larkfinserv-react-native-sdk/src/types/index.ts
export type EligibilityResult = {
  eligible: boolean;
  amount?: number;
  interestRate?: number;
  terms?: string[];
  reasons?: string[];
  mobile?: string;
  partnerId?: string;
  sessionId?: string;
  result?: {
    selectedOffer: any;
  };
};

export interface ThemeConfig {
  primaryColor?: string;
  secondaryColor?: string;
  fontFamily?: string;
  logoUrl?: string;
  name?: string;
}

export interface ConsentData {
  privacyConsent: boolean;
  promotionalConsent: boolean;
}

export interface StepProgressSummary {
  totalSteps: number;
  completedSteps: number;
  currentStep: string | null;
  nextStep: string | null;
  completionPercentage: number;
}

export interface LoanDetails {
  loanAccountId: string;
  loanStatus: string;
  sanctionAmount: number;
  totalOutstandingAmount: number;
  availableAmountForDisbursement: number;
  principalOutstandingAmount: number;
  interestRate: number;
  tenureInDays: number;
  currentTermStartDate: number;
  currentTermEndDate: number;
}

export interface PartnerConfig {
  partnerId?: string;
  partnerName?: string;
  userId?: string;
  userData?: Record<string, unknown>;
  sessionId?: string;
  phoneNumber?: string;
  consentData?: ConsentData;
  theme?: ThemeConfig;
  apiKey: string;
  apiSecret: string;
  environment?: 'sandbox' | 'production';
  currentJourneyStep?: string;
  stepProgressSummary?: StepProgressSummary;
  hasActiveLoan?: boolean;
  loanDetails?: LoanDetails;
  requiresEmailVerification?: boolean;
}

export type SDKEventType =
  | 'READY'
  | 'ELIGIBILITY_RESULT'
  | 'ERROR'
  | 'CLOSE'
  | 'INITIATED'
  | 'CLOSE_FRAME'
  | 'JOURNEY_STEP_COMPLETED'
  | 'SESSION_CLEARED'
  | 'SDK_REINITIALIZED'
  | 'REINITIALIZATION_FAILED';

export type SDKMode = 'popup' | 'inline' | 'embedded';

export interface SDKEventData {
  sessionId?: string;
  themeConfig?: Record<string, unknown>;
  error?: SDKError;
  result?: {
    status: string;
    data: Record<string, unknown>;
  };
  completedStep?: string;
  nextStep?: string;
  stepSummary?: StepProgressSummary;
  reason?: string;
  mobileNumber?: string;
  timestamp?: string;
  [key: string]: any;
}

export interface SDKEvent {
  type: SDKEventType;
  data?: SDKEventData;
}

export interface SDKError {
  code: string;
  message: string;
}

export type EventHandler = (event: SDKEvent) => void;
